Don't Store WordPress.com Passwords

jleandroperez opened the pull request:
### Description:
In this Pull Request, we enhance the ViewControllers outlined below, so that whenever a WordPress.com URL needs authentication, we send over the *Bearer Token* via the HTTP headers.

This enables us to stop storing the WordPress.com password in the keychain, and also allows us to prevent login errors due to Multifactor-Enabled accounts: the app might not know the latest MFA code, but the OAuth token is expected to be valid.

Plus, **WPComOAuthController** has been enhanced, so the Single Sign On mechanism also relies on the Bearer Token (scenario outlined in Testing Flow F).

Closes #1220

#### Enhanced viewControllers:
- **WPWebViewController**
- **PostPreviewViewController**
- **StatsWebViewController**

-- 

### Testing Scenarios:
Below, we'll detail several testing flows that have been considered while implementing this enhancement.

#### Flow A: Private Blog
1. Set a dotcom 'Testing Blog' as private
2. Open My Sites > 'Testing Blog'
3. Tap over 'View Site'

*Expected Results:*
- The site is visible, and WordPress.com doesn't ask for a password

<br/>

#### Flow B: Web Stats
1. Open My Sites > 'Testing Blog'
2. Tap over 'Stats'
3. Scroll down, and tap over 'View Web Version'

*Expected Results:*
- The stats are visible, and WordPress.com doesn't ask for a password

<br/>

#### Flow C: Post Preview
1. Open My Sites > 'Testing Blog'
2. Tap over Blog Posts
3. Tap over any post
4. Press the right top eye button

*Expected Results:*
- The post is visible, and WordPress.com doesn't ask for a password

<br/>

#### Flow D: Edit Site
1. Fresh Install
2. Log into a self hosted 'Blog A'
3. Install 'Disable XML-RPC' plugin
4. Open wp-admin in Safari, and update your dotcom password
5. Open 'My Sites' tab
6. Tap over 'Blog A' row
7. Tap 'Settings'
8. Update your password.

*Expected Results:*
- The Self Hosted blog is expected to return an error with code 405 (XMLRPC Disabled). 
- The app should push a WebView with your wp-admin already logged in.

<br/>

#### Flow E: Install jetpack
1. Open wp-admin in Safari, and turn off Jetpack in your self hosted blog.
2. Log into your self hosted site in WPiOS
3. Tap over 'My Sites' > 'Self Hosted Blog'
4. Tap over 'Stats'
5. Tap over 'Install Jetpack'

*Expected Results:*
- The app should push a WebView with your Self Hosted blog already logged in, and the **Install Jetpack View** onscreen.

<br/>

#### Flow F: Single Sign On
1. Install WPiOS and Sign Into a dotcom account.
2. Add a [new testing WordPress App in the Developer Portal](https://developer.wordpress.com/apps/)
3. Open this url in Safari.app: `wordpress-oauth-v2://authorize?client_id=[App ID]&redirect_url=[Callback URL]`

*Expected Results:*
- WPiOS should be onscreen, and the user should be asked whether if he wants to authorize or not the app. 
- No login or password should be required.

jleandroperez commented:
Since this is *a very* delicate change, i'm hereby invoking ~~the Klingon High Council~~ the Architecture High Council for a review.

@aerych @astralbodies @sendhil 

Thanks in advance gentleman. Testing flows outlined in the PR, if i did miss something (or anything), please, let me know.

sendhil commented:
Oh boy - that's a big change @jleandroperez  ;). I'll try to peek at this tomorrow...

sendhil commented:
cc @diegoreymendez  for good measure

aerych commented on WordPress/Classes/ViewRelated/Reader/WPWebAppViewController.m:
```diff
@@ -1,5 +1,5 @@
 #import "WPWebAppViewController.h"
```
It makes me just a little sad you had to update this one. If it weren't for friendfinder it could just go away :(

aerych commented on WordPress/Classes/ViewRelated/Reader/WPWebBridge.m:
```diff
@@ -1,6 +1,6 @@
 #import "WPWebBridge.h"
```
This one too :/

aerych commented:
Quite the PR @jleandroperez :)  I promise to spend some time trying to violently break it. 

I spent some time looking through the code changes. Something I noticed is it removes the code in the app delegate that sets the default app user agent (assigns a value to the NSUserDefaults key UserAgent). This is particular important for UIWebViews so all content is requested with a uniform UA. With this change, you might specify the app's UA in the NSURLRequest passed to a UIWebView, but any asset request for the loaded page (js, css, images, etc) will use the default iOS UA. 

For tracking purposes, the webviews used by the reader and stats should make all requests with the app UA. The web view used to show a post preview should use the default iOS UA for its requests (which fixes an issue with how some themes rendered the preview based on UA). 

Can we restore this functionality?

aerych commented on WordPress/Classes/System/WordPressAppDelegate.m:
```diff
@@ -836,60 +835,18 @@ - (void)cleanUnusedMediaFileFromTmpDir
     }];
 }
 
-#pragma mark - Networking setup, User agents
```
Just following up to my comment on the PR. We should find a way to keep the functionality in `setupUserAgent`, `useDefaultUserAgent` and `useAppUserAgent`.

jleandroperez replied:
(Already restored, thanks for the heads up Eric!)

jleandroperez commented:
@aerych i've replaced the User-Agent getters / logic with something simpler:

"Default user agent" is the one that's sent if you don't specify anything different, so i just added `[[UIDevice currentDevice] wordPressUserAgent]`

*However*, i was not aware that UIWebView actually picks up the User-Agent from NSUserDefaults (i updated all of the *NSURLRequests*). Meaning that, indeed, you're right. I'm restoring those helpers, thanks for the heads up Eric!

jleandroperez commented:
@aerych functionality restored, and ready for the fire test. Thank you sir!

P.s.: I would love to nuke... WPWebAppViewController + WPWebBridge + WPChromelessWebViewController.

But perhaps that deserves another issue (it's a bad idea to break multiple things in just one PR!).

aerych commented:
> UIWebView actually picks up the User-Agent from NSUserDefaults

Yah that was really all I was trying to say :) I used too many words. ;)

> [...] perhaps that deserves another issue

For sure!

diegoreymendez commented on WordPress/Classes/Categories/NSURLRequest+Helpers.m:
```diff
@@ -0,0 +1,61 @@
+    NSParameterAssert(username);
+    NSParameterAssert(password != nil || bearerToken != nil);
+    
+    NSMutableURLRequest *request = [[self requestWithURL:loginUrl userAgent:userAgent] mutableCopy];
```
Why is it necessary to copy the request object here?

jleandroperez replied:
You're indeed right, i meant to change that down the road, but obviously forgot to address it. Will update soon!

diegoreymendez replied:
:+1:

diegoreymendez commented on WordPress/Classes/Categories/NSURLRequest+Helpers.h:
```diff
@@ -0,0 +1,14 @@
+#import <Foundation/Foundation.h>
+
+@interface NSURLRequest (Helpers)
```
This is pretty minor, but it may make more sense to create a subclass (such as `WPURLRequest`) instead of a category, since the code in `requestForAuthenticationWithURL` seems to be WP-specific / non-generic.

jleandroperez replied:
Perhaps we could rename the getter to.. `requestForWordPressAuthentication`?. 

I'm not comfortable with the idea of subclassing NSURLRequest, mostly because we wouldn't be adding new attributes to it. This new class would contain a couple static getters (that could return just a NSURLRequest instance).

diegoreymendez replied:
So I'll explain the reasoning behind my comment, but keep in mind the current implementation is OK and you can opt to keep it.

So the reason why I consider subclassing a better option here is because of single-responsibility.  The class `NSURLRequest` has the single responsibility of providing us with a generic interface to make URL requests, without caring about what those requests are used for.  By adding a method to that class that knows what our requests look like, it feels to me like we're stepping a bit outside of that class' domain.

However, with that in mind, feel free to leave it as it is.  Just my 2 cents.

jleandroperez replied:
You're indeed right Diego, and you also do have a great taste. Let's go for it, thank you!

diegoreymendez commented on WordPress/Classes/Categories/NSURLRequest+Helpers.h:
```diff
@@ -0,0 +1,14 @@
+
++ (NSURLRequest *)requestWithURL:(NSURL *)url userAgent:(NSString *)userAgent;
+
++ (NSURLRequest *)requestForAuthenticationWithURL:(NSURL *)loginUrl
```
I'd consider documenting this class and these methods since they may be non-obvious to someone peeking through the code.

jleandroperez replied:
Good call, adding!

diegoreymendez commented on WordPress/Classes/Models/WPAccount.m:
```diff
@@ -52,6 +52,10 @@ - (void)didTurnIntoFault
 - (NSString *)password
 {
+    if (self.isWpcom) {
+        return [NSString string];
```
I was trying to figure out why we need to return an empty `NSString` in this case, but couldn't really figure it out.  Can you clarify?

My main concern is that internally we tend to use `nil` for unset properties - and this would go out of the norm a bit, which I would avoid if possible.  If some interface needs a string instead of a nil value, I believe it should be that interface to make the appropriate conversion.  Let me know.

jleandroperez replied:
I took the empty string road, there, mostly, [to avoid this kind of scenario](https://github.com/wordpress-mobile/WordPress-iOS/blob/develop/WordPress/Classes/Networking/PostServiceRemoteXMLRPC.m#L129) (which would otherwise require us to check case by case).

However, on second thoughts, XMLRPC shouldn't be called for dotcom accounts. I'll make the arrangements so the password is *nil* (and will also add failsafes where needed).

Thank you Diego!

aerych commented:
I might be missing it in all the changes but I haven't found where an existing wpcom is removed from the keychain. Is it in there? If not, should it be added?

diegoreymendez commented:
@lantean - It seems oclint is failing.  I'm copying the logs here for ease of reading:

```
/Users/travis/build/wordpress-mobile/WordPress-iOS/WordPress/Classes/Models/Blog.m:286:9: collapsible if statements P3 
/Users/travis/build/wordpress-mobile/WordPress-iOS/WordPress/Classes/Models/Blog.m:67:5: ivar assignment outside accessors or init P2 
/Users/travis/build/wordpress-mobile/WordPress-iOS/WordPress/Classes/Models/Blog.m:68:5: ivar assignment outside accessors or init P2 
/Users/travis/build/wordpress-mobile/WordPress-iOS/WordPress/Classes/Models/Blog.m:301:6: ivar assignment outside accessors or init P2 
/Users/travis/build/wordpress-mobile/WordPress-iOS/WordPress/Classes/Models/Blog.m:304:6: ivar assignment outside accessors or init P2 
/Users/travis/build/wordpress-mobile/WordPress-iOS/WordPress/Classes/Models/Blog.m:16:1: too many methods P3 Objective-C implementation with 32 methods exceeds limit of 30
/Users/travis/build/wordpress-mobile/WordPress-iOS/WordPress/Classes/Networking/PostServiceRemoteXMLRPC.m:111:1: high cyclomatic complexity P2 Cyclomatic Complexity Number 11 exceeds limit of 10
/Users/travis/build/wordpress-mobile/WordPress-iOS/WordPress/Classes/Networking/PostServiceRemoteXMLRPC.m:231:1: high cyclomatic complexity P2 Cyclomatic Complexity Number 12 exceeds limit of 10
/Users/travis/build/wordpress-mobile/WordPress-iOS/WordPress/Classes/Networking/PostServiceRemoteXMLRPC.m:231:1: high ncss method P2 Method of 35 non-commenting source statements exceeds limit of 30
/Users/travis/build/wordpress-mobile/WordPress-iOS/WordPress/Classes/Networking/PostServiceRemoteXMLRPC.m:231:1: high npath complexity P2 NPath Complexity Number 320 exceeds limit of 200
/Users/travis/build/wordpress-mobile/WordPress-iOS/WordPress/Classes/Networking/PostServiceRemoteXMLRPC.m:272:9: replace with object subscripting P3 
/Users/travis/build/wordpress-mobile/WordPress-iOS/WordPress/Classes/Networking/PostServiceRemoteXMLRPC.m:276:9: replace with object subscripting P3 
/Users/travis/build/wordpress-mobile/WordPress-iOS/WordPress/Classes/Networking/PostServiceRemoteXMLRPC.m:284:9: replace with object subscripting P3 
/Users/travis/build/wordpress-mobile/WordPress-iOS/WordPress/Classes/Networking/PostServiceRemoteXMLRPC.m:286:5: replace with object subscripting P3 
/Users/travis/build/wordpress-mobile/WordPress-iOS/WordPress/Classes/Models/WPAccount.m:47:5: ivar assignment outside accessors or init P2 
/Users/travis/build/wordpress-mobile/WordPress-iOS/WordPress/Classes/Models/WPAccount.m:48:5: ivar assignment outside accessors or init P2 
```

jleandroperez commented:
@diegoreymendez P2 error refers to cyclomatic complexity // code improvements. None of those errors apply to the changes introduced in this PR:

- *Blog.m* wasn't modified, in the lines mentioned (Affected ones are 310-311 344-348). 
- *PostServiceRemoteXMLRPC.m*, it just got few extra *NSParameterAssert*'s (116-119).
- *WPAccount.m* got lines 55-58 modified.

Do you mind if we file another issue to address those warnings?

@aerych good catch Eric! i'll add a helper to nuke existing dotcom passwords.

jleandroperez commented:
@aerych 'Nuke Dotcom Password' mechanism added, and ready for review. Thank you!

aerych commented:
I walked through the testing scenarios in the simulator except for D and E -- I'm not currently equipped to test those. 
A - C worked flawlessly. They just wouldn't break for me. Props!
F opened the app for me but I didn't see a prompt.  I tried with two different apps, one configured for web and one configured for native. In both cases the callback URLs were not longer accessible, but that shouldn't have prevented a prompt, and I would have expected an error message in that case regardless.  Not sure if the simulator is cranky or if something is amiss.

aerych commented:
I've gone through the code twice now. Apart from @diegoreymendez's comments (which you've already addressed) I have nothing to add. :)

jleandroperez commented:
@aerych thank you very much!. This URI should do the trick:

`wordpress-oauth-v2://authorize?client_id=39580&redirect_uri=http://www.lantean.co`

I've just noticed that, as a side effect of one of the latest changes (stop returning nil passwords), the SSO helper breaks. Will get it fixed asap, thanks Eric!

diegoreymendez commented:
@jleandroperez - Opening another issue to review the logs I posted it totally acceptable - In fact I just opened a separate ticket and assigned it to myself [here](https://github.com/wordpress-mobile/WordPress-iOS/issues/3340).

jleandroperez commented:
@diegoreymendez thanks for adding that one!. I'm adding it to my list as well, whoever gets loose first picks it up!.

@aerych WordPressApi dependency updated. Opening this link should help you test Scenario F:

`wordpress-oauth-v2://authorize?client_id=39580&redirect_uri=http://www.lantean.co`

If there's anything else we should update / fix / tweak, please, let me know =)

Thank you all!

aerych commented:
Took Scenario F for another spin.  Tested with the link you provided @jleandroperez and also with a link for one of my test apps.  Both worked perfectly this time. 
Nice :)

aerych commented:
I would like to confirm that Scenarios D and E have been thoroughly tested, but otherwise this patch has a :+1: from me.

jleandroperez commented:
@aerych thank you very much sir!

aerych commented:
Took Scenarios D and E for a test drive (thanks for the test credentials @jleandroperez!) and everything worked as expected. 
:shipit: from me :)

jleandroperez commented:
@aerych *THANK YOU*!!!
